pkg/crypto/certloader/watcher.go

```diff
@@ -0,0 +1,239 @@
+	if err != nil {
+		return nil, err
+	}
+	w := &Watcher{
+		FileReloader: r,
+		log:          log,
+		fswatcher:    fswatcher,
+		stop:         make(chan struct{}),
+	}
+
+	res := make(chan *Watcher)
+	go func(res chan<- *Watcher) {
+		defer close(res)
+		// Attempt a reload without having received any fs notification in case
+		// all the files are already there. Note that the keypair and CA are
+		// reloaded separately as a "partial update" is still useful: If the
+		// FileReloader is "half-ready" (e.g. has loaded the keypair but failed
+		// to load the CA), we only need a successfully handled CA related fs
+		// notify event to become Ready (in other words, we don't need to
+		// receive a fs event for the keypair in that case to become ready).
+		_, keypairErr := w.ReloadKeypair()
+		_, caErr := w.ReloadCertificateAuthority()
+		ready := w.Watch()
+		if keypairErr == nil && caErr == nil {
+			log.Debug("TLS configuration ready")
+			res <- w
+			return
+		}
+		log.Debug("waiting on fsnotify update to be ready")
+		select {
+		case <-ready:
+			log.Debug("TLS configuration ready")
+			res <- w
+		case <-w.stop:
+		}
+	}(res)
+
+	return res, nil
+}
+
+// Watch initialize the files watcher and update goroutine. It returns a ready
+// channel that will be closed once an update made the underlying FileReloader
+// ready.
+func (w *Watcher) Watch() chan struct{} {
+	// build maps for the CA files and keypair files to help detecting what has
+	// changed in order to reload only the appropriate certificates.
+	keypairMap := make(map[string]struct{})
+	caMap := make(map[string]struct{})
+	if w.FileReloader.certFile != "" {
+		keypairMap[w.FileReloader.certFile] = struct{}{}
+	}
+	if w.FileReloader.privkeyFile != "" {
+		keypairMap[w.FileReloader.privkeyFile] = struct{}{}
+	}
+	for _, path := range w.FileReloader.caFiles {
+		caMap[path] = struct{}{}
+	}
+
+	// prepare the ready channel to be returned. We will close it exactly once.
+	var once sync.Once
+	ready := make(chan struct{})
+	go func() {
+		defer w.fswatcher.Close()
+		for {
+			select {
+			case event := <-w.fswatcher.Events:
+				path := event.Name
+				log := w.log.WithFields(logrus.Fields{
+					logfields.Path: path,
+					"operation":    event.Op,
+				})
+				log.Debug("Received fsnotify event")
+				_, keypairChanged := keypairMap[path]
+				_, caChanged := caMap[path]
+				createOperation := event.Op&fsnotify.Create == fsnotify.Create
+				if (keypairChanged || caChanged) && createOperation {
+					// A file we are interested in has been created. We might
+					// have been watching only over its directory until now,
+					// add it to the watcher since it exists.
+					if err := w.fswatcher.Add(path); err != nil {
+						log.WithError(err).Warnf("Failed to add %s to fsnotify watcher", logfields.Path)
+					}
+				}
+				if keypairChanged {
+					keypair, err := w.ReloadKeypair()
+					if err != nil {
+						log.WithError(err).Warn("keypair update failed")
+						continue
```

Comment:
I think it would be useful to add a `Err() error` (or `LastErr() error` ?) method that returns the last error encountered (or maybe even add a chan of errors). This would allow the caller to decide to keep waiting or bail out depending on whether the error is transient or not. Right now, the error is simply logged and discarded which makes it impossible to handle by the caller.

Reply:
I'm not found of the `Err()` or `LastErr()` method as it would require locking.

An error channel could be interesting. We could easily add something like:
```go
// used by the caller
func (w *Watcher) Errors() <-chan error {
    return w.errors
}

// used by Watch()
func (w *Watcher) reportWatchError(err error) {
    select {
    case w.errors <- err:
    default:
        w.log.WithError(err).Warn("Unhandled watcher error")
    }
}
```
Then, I think we would have to create different errors depending how the caller want to react to them.

Now, in both hubble server and relay, the caller would only log the error and go on so I think we'll keep it as it stand for now.